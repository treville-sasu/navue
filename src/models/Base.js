export class Model {
  // eslint-disable-next-line no-unused-vars
  constructor({ type, ...properties } = {}) {
    Object.assign(this, properties);
  }

  toJSON() {
    return {
      ...this,
      type: this.constructor.name
    };
  }

  // TODO write specs
  clone() {
    return this.constructor.from(JSON.parse(JSON.stringify(this)));
  }

  static from({ type, ...properties } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type == this.name) return new this(properties);
    else throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
  }
}

// TODO : implementation for when Vuejs Suppots Array-like objects.
// export class Store extends Array {
export class Store {
  constructor(properties = {}, ...items) {
    this.items = items;
    Object.assign(this, properties);
  }

  // get items() {
  //   return [...this];
  // }

  [Symbol.iterator]() {
    return this.items[Symbol.iterator]();
  }

  get length() {
    return this.items.length;
  }

  valueOf() {
    return this.items;
  }

  get properties() {
    // eslint-disable-next-line no-unused-vars
    let { items, ...props } = { ...this };
    return props;
    // return Object.fromEntries(
    //   Object.entries(this).filter(([key]) => isNaN(key))
    // );
  }

  get first() {
    return this.items[0];
  }

  get last() {
    return this.items[this.length - 1];
  }

  add(item, index) {
    index ? this.items.splice(index, 0, item) : this.items.push(item);
    return this.length;
  }

  remove(item, index) {
    if (index === undefined) index = this.items.indexOf(item);
    return index >= 0 ? this.items.splice(index, 1).length == 1 : false;
  }

  keep(filter) {
    let i = this.items.length;
    while (i--) {
      let result = filter.call(this.items, this.items[i]);

      if (result == undefined) this.items.splice(i, 1);
      else if (!Object.is(result, this.items[i]))
        this.items.splice(i, 1, result);
    }
    return this;
  }

  toJSON() {
    return {
      ...this.properties,
      items: this.items,
      type: this.constructor.name
    };
  }

  static from({ type, items = [], ...properties }, ...constructors) {
    if (type != this.name)
      throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
    else {
      let constructor = constructors.shift();
      return new this(
        properties,
        ...items.map(i => {
          return constructor ? constructor.from(i, ...constructors) : i;
        })
      );
    }
  }
}
