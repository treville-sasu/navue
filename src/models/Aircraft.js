import { Model, Store } from "@/models/Base.js";
import { Speed, Consumption, Weight, Distance } from "@/models/Quantities.js";

export class Aircraft extends Model {
  constructor({
    registration,
    manufacturer,
    model,
    paces = new Store(),
    consumptions = new Store(),
    balance = new Store({ date: undefined }),
    envelopes = new Store(),
    checklists = new Store()
  } = {}) {
    paces.keep(i => (i instanceof Speed ? i : undefined));
    consumptions.keep(i => (i instanceof Consumption ? i : undefined));
    balance.keep(i => (i instanceof Weight ? i : undefined));
    envelopes.keep(i => (i instanceof Store ? i : undefined));
    // envelopes = envelopes.keep(i => {
    //   return i.keep(j => (j instanceof Weight ? j : undefined));
    // });
    checklists.keep(i => (i instanceof Store ? i : undefined));

    super({
      registration,
      manufacturer,
      model,
      paces,
      consumptions,
      balance,
      envelopes,
      checklists
    });
  }

  static from({
    _id,
    _rev,
    type,
    registration,
    manufacturer,
    model,
    paces,
    consumptions,
    balance,
    envelopes,
    checklists
  } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type != this.name)
      throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
    else {
      if (paces) paces = Store.from(paces, Speed);
      if (consumptions) consumptions = Store.from(consumptions, Consumption);
      if (balance) balance = Store.from(balance, Weight);
      if (envelopes) envelopes = Store.from(envelopes, Store, Moment);
      if (checklists) checklists = Store.from(checklists, Store);

      let imported = new this({
        registration,
        manufacturer,
        model,
        paces,
        consumptions,
        balance,
        envelopes,
        checklists
      });
      if (_id && _rev) Object.assign(imported, { _id, _rev });

      return imported;
    }
  }
}

export class Moment extends Model {
  constructor(mass, lever, properties) {
    if (!mass) mass = new Weight();
    if (!lever) lever = new Distance();

    super({ mass, lever, ...properties });
  }

  get value() {
    return this.mass.value * this.lever.value;
  }

  get _value() {
    return this.mass * this.lever;
  }

  get unit() {
    return `${this.mass.unit}.${this.lever.unit}`;
  }

  valueOf() {
    return this._value;
  }

  toJSON() {
    return {
      ...this,
      value: this.value,
      unit: this.unit,
      type: this.constructor.name
    };
  }
  toCoords() {
    return {
      x: this.lever.value,
      y: this.mass.value
    };
  }

  static from({ type, mass, lever } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type == this.name) {
      if (mass) mass = Weight.from(mass);
      if (lever) lever = Distance.from(lever);
      return new this(mass, lever, { type });
    } else throw `Supplied object is no ${this.name}`;
  }
}
