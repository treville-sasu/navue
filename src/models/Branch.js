import { Model } from "@/models/Base.js";
import { Distance } from "@/models/Quantities.js";

import { lineString, featureCollection, round } from "@turf/helpers";
import bbox from "@turf/bbox";
import length from "@turf/length";
import buffer from "@turf/buffer";
import truncate from "@turf/truncate";

export class Branch extends Model {
  constructor(member, properties = {}, ...features) {
    if (member)
      features.forEach((f, i, arr) => {
        arr[i] = member.from(f);
      });
    super({ ...featureCollection(features), properties, member });
  }

  first() {
    return this.features[0];
  }

  last() {
    return this.features[this.features.length - 1];
  }

  push(...positions) {
    return this.features.push(
      ...positions.map(p => (this.member ? this.member.from(p) : p))
    );
  }

  shift() {
    return this.features.shift();
  }

  insert(position, index) {
    const pos = isNaN(index) ? this.features.length : index;
    this.features.splice(pos, 0, position);
    return position;
  }

  remove(index) {
    this.features.splice(index, 1);
    return this.features.length;
  }

  forEach(cb) {
    return this.features.forEach(cb);
  }

  map(cb) {
    return this.features.map(cb);
  }

  filter(cb) {
    return this.features.filter(cb);
  }

  pairs(cb) {
    let collection = [];
    if (this.features.length)
      this.features.reduce((previous, current, i, a) => {
        if (previous) collection.push(cb.call(this, previous, current, i, a));
        return current;
      });
    return collection;
  }

  get geom() {
    return this.map(({ geometry: { coordinates } }) => coordinates);
  }

  get bbox() {
    return bbox(lineString(this.geom));
  }

  get length() {
    return new Distance(
      round(length(lineString(this.geom), { units: "meters" }))
    );
  }

  toJSON() {
    // eslint-disable-next-line no-unused-vars
    let { member, ...json } = super.toJSON();
    return json;
  }

  toGeoJSON(type) {
    switch (type) {
      case "Feature":
      case "LineString":
        return lineString(this.geom, this.properties, { bbox: this.bbox });
      case "MultiPolygon":
        return featureCollection(
          this.pairs((start, end) => {
            return truncate(
              buffer(
                // TODO : Should we create midpoints and average the properties ? let's see in real life
                lineString(
                  [start.geometry.coordinates, end.geometry.coordinates],
                  start.properties
                ),
                Number(start.properties.accuracy) || 15,
                {
                  units: "meters",
                  steps: 1
                }
              ),
              { mutate: true }
            );
          }),
          { bbox: this.bbox }
        );
      case "FeatureCollection":
      default:
        return {
          ...featureCollection(this.features, {
            bbox: this.bbox
          }),
          properties: this.properties
        };
    }
  }

  static from({ type, features = [], properties }, constructor) {
    if (arguments[0] instanceof this) return arguments[0];

    if (type === this.name || type === "FeatureCollection")
      return new this(constructor, properties, ...features);
    else
      throw `Invalid data : 'type' should be '${this.name}' or 'FeatureCollection' got '${type}'`;
  }
}
