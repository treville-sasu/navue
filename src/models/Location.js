import { Waypoint } from "@/models/Waypoint.js";
import { Altitude, Distance, Speed, Azimuth } from "@/models/Quantities.js";

import bbox from "@turf/bbox";
import buffer from "@turf/buffer";

export class Location extends Waypoint {
  get bbox() {
    return bbox(
      buffer(this.toGeoJSON(), Number(this.properties.accuracy), {
        units: "meters"
      })
    );
  }

  bounds(radius) {
    return bbox(
      buffer(this.toGeoJSON(), radius, {
        units: "meters"
      })
    );
  }

  legFrom(other) {
    return {
      distance: other.distanceTo(this),
      heading: other.bearingTo(this),
      speed: this._speedFrom(other),
      verticalSpeed: this._verticalSpeedFrom(other)
    };
  }

  willBeIn(delay) {
    if (this.properties.speed && this.properties.heading)
      return this.destinationPoint(
        this.properties.speed * delay,
        this.properties.heading
      );
    else
      throw "cannot calculate futur position, speed & heading should be provided";
  }

  _delayFrom({ properties: { timestamp } }) {
    return (this.properties.timestamp - timestamp) / 1000;
  }

  _climbFrom({ properties: { altitude } }) {
    return this.properties.altitude - altitude;
  }

  _speedFrom(other) {
    if (this._delayFrom(other) == 0)
      throw "Teleportation is not yet a possibility";
    return new Speed(this.distanceTo(other) / this._delayFrom(other));
  }

  _verticalSpeedFrom(other) {
    if (this._delayFrom(other) == 0)
      throw "Teleportation is not yet a possibility";

    return new Speed(this._climbFrom(other) / this._delayFrom(other));
  }

  static from({ accuracy, altitude, speed, heading, ...others } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    if (accuracy) accuracy = Distance.from(accuracy);
    if (altitude) altitude = Altitude.from(altitude);
    if (speed) speed = Speed.from(speed);
    if (heading) heading = Azimuth.from(heading);
    return super.from({ accuracy, altitude, speed, heading, ...others });
  }

  static fromGeolocationPosition({
    coords: {
      latitude,
      longitude,
      altitude,
      accuracy,
      altitudeAccuracy,
      heading,
      speed
    },
    timestamp
  }) {
    if (altitude)
      altitude = new Altitude(altitude, "m", {
        reference: "WGS84",
        accuracy: altitudeAccuracy
      });

    if (accuracy) accuracy = new Distance(accuracy, "m");
    if (speed) speed = new Speed(speed, "m/s");
    if (heading) heading = new Azimuth(heading, "Â°");

    return new this([longitude, latitude, altitude], {
      accuracy,
      altitude,
      speed,
      heading,
      timestamp
    });
  }
}
