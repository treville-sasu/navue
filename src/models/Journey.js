import { Model } from "@/models/Base";
import { Branch } from "@/models/Branch";
import { Distance } from "@/models/Quantities";

import { multiLineString, featureCollection, round } from "@turf/helpers";
import bbox from "@turf/bbox";
import length from "@turf/length";

export { Branch };
// FIXME: create branche if none exists
export class Journey extends Model {
  constructor(properties = {}, ...branches) {
    super({ properties, branches });
  }

  get name() {
    return this.properties.name;
  }

  set name(val) {
    return (this.properties.name = val);
  }

  last() {
    return this.branches[this.branches.length - 1];
  }

  push(...positions) {
    return this.last().push(...positions);
  }

  shift() {
    return this.last().shift();
  }

  forEach(cb) {
    return this.branches.forEach(cb);
  }

  map(cb) {
    return this.branches.map(cb);
  }

  filter(cb) {
    return this.branches.filter(cb);
  }

  addBranch(locations = [], properties) {
    this.branches.push(new Branch(undefined, properties, ...locations));
    return this;
  }

  getBranch(ref) {
    if (ref instanceof Branch) return ref;
    else if (Number.isInteger(ref)) return this.branches[ref];
    else return undefined;
  }

  removeBranch(ref) {
    let index = this.branches.indexOf(this.getBranch(ref));
    this.branches.splice(index, 1);
    return this;
  }

  get geom() {
    return this.map(b => b.geom);
  }

  get bbox() {
    return bbox(multiLineString(this.geom));
  }

  get length() {
    return new Distance(
      round(length(multiLineString(this.geom), { units: "meters" }))
    );
  }

  //TODO Specs multi types
  toGeoJSON(...types) {
    if (types.length > 1) {
      return featureCollection(
        types.flatMap(type => {
          let feat = this.toGeoJSON(type);
          return feat.features ? feat.features : feat;
        })
      );
    } else
      switch (types[0]) {
        case "Feature":
        case "LineString":
          return featureCollection(
            this.branches.flatMap(b => b.toGeoJSON("LineString"), {
              bbox: this.bbox
            })
          );
        case "MultiLineString":
          return featureCollection(
            this.branches.flatMap(
              b => b.toGeoJSON("MultiLineString").features,
              {
                bbox: this.bbox
              }
            )
          );
        case "FeatureCollection":
        default:
          return featureCollection(
            this.branches.flatMap(
              b => b.toGeoJSON("FeatureCollection").features
            ),
            { bbox: this.bbox }
          );
      }
  }

  static from({ type, branches = [], properties, ...others }) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type === this.name) {
      let created = new this(properties);
      branches.forEach(({ features, properties }) => {
        created.addBranch(features, properties);
      });

      Object.assign(created, others); //allow for private props like _id, _rev....
      return created;
    } else throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
  }
}
