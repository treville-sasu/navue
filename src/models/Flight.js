import { Journey } from "@/models/Journey";
import { Location } from "@/models/Location";
import { Branch } from "./Journey";

export class Flight extends Journey {
  constructor(
    { name, notes, checked = [], milestones = [] } = {},
    ...locations
  ) {
    // let branches;
    // if (locations.every(a => a instanceof Location))
    //   branches = [new Branch(Location, {}, ...locations)];
    // else if (
    //   locations.every(
    //     a => a instanceof Array && a.every(b => b instanceof Branch)
    //   )
    // )
    //   branches = locations.flat();
    let branches = locations;
    super({ name, notes, checked, milestones }, ...branches);
  }

  addBranch(locations, properties) {
    this.branches.push(new Branch(Location, properties, ...locations));
    return this;
  }

  get duration() {
    return (
      this.branches.reduce((total, b) => {
        return (total +=
          b.last().properties.timestamp - b.first().properties.timestamp);
      }, 0) / 1000
    );
  }

  // static from({ type, locations, features, ...properties } = {}) {
  //   if (arguments[0] instanceof this) return arguments[0];

  //   switch (type) {
  //     case this.name:
  //       if (locations)
  //         locations = locations.map(t => t.map(l => Location.from(l)));

  //       break;
  //     case "FeatureCollection":
  //       locations = [features.map(l => Location.from(l))];
  //       // properties = { ...properties };
  //       break;

  //     default:
  //       throw `Invalid data : 'type' should be 'FeatureCollection' or '${this.name}' got '${type}'`;
  //   }

  //   return new this(properties, ...locations);
  // }
}
