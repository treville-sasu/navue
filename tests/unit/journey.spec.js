import { Branch, Journey } from "@/models/Journey";
import { Model } from "@/models/Base";
import { Location } from "@/models/Location";

describe("branch", () => {
  let empty, numbered, branch;
  beforeEach(() => {
    empty = new Branch();
    numbered = new Branch(Number, { a: 1 }, 1, 2);

    class Position {
      constructor(args) {
        Object.assign(this, args);
      }
      static from(args) {
        return new this(args);
      }
    }

    branch = new Branch(
      Position,
      { a: 1 },
      {
        geometry: {
          coordinates: [0, 0, 1],
          type: "Point"
        },
        properties: {
          altitude: 1
        },
        type: "Feature"
      },
      {
        geometry: {
          coordinates: [1, 1, 0],
          type: "Point"
        },
        properties: {
          altitude: 0
        },
        type: "Feature"
      },
      {
        geometry: {
          coordinates: [-1, -1, 2],
          type: "Point"
        },
        properties: {
          altitude: 0
        },
        type: "Feature"
      }
    );
  });

  it("extends Model", () => {
    expect(Branch.prototype).toBeInstanceOf(Model);
  });

  // it.skip("is registerd globaly", () => {
  //   expect(classes["Branch"]).toBe(Branch);
  // });

  it("is instantiated", () => {
    expect(empty).toHaveProperty("features", []);
    expect(empty).toHaveProperty("member", undefined);
    expect(empty).toHaveProperty("properties", {});

    expect(numbered).toHaveProperty("member", Number);
    expect(numbered).toHaveProperty("features", [1, 2]);
    expect(numbered).toHaveProperty("properties", { a: 1 });
  });

  it("hold, geom, bbox & length", () => {
    expect(branch.geom).toMatchSnapshot();
    expect(branch.bbox).toMatchSnapshot();
    expect(branch.length).toMatchSnapshot();
  });

  it("should behave a bit like an array", () => {
    let features;
    // FIXME: member.from should be replaced by new Member()
    features = jest.spyOn(empty.features, "push");
    empty.push(1, 2);
    expect(features).toHaveBeenCalled();

    features = jest.spyOn(branch.features, "push");
    branch.push(1, 2);
    expect(features).toHaveBeenCalled();
    features = jest.spyOn(branch.features, "shift");
    branch.shift(1, 2);
    expect(features).toHaveBeenCalled();

    features = jest.spyOn(branch.features, "forEach");
    branch.forEach(x => x);
    expect(features).toHaveBeenCalled();

    features = jest.spyOn(branch.features, "map");
    branch.map(x => x);
    expect(features).toHaveBeenCalled();

    features = jest.spyOn(branch.features, "filter");
    branch.filter(() => true);
    expect(features).toHaveBeenCalled();
  });

  it.todo("insert new position at given index");
  it.todo("remove position at given index");

  it("should return first & last feature", () => {
    expect(branch.first()).toBe(branch.features[0]);
    expect(branch.last()).toBe(branch.features[branch.features.length - 1]);
  });

  it("should iterate over pairs", () => {
    expect(branch.pairs((p, c, i) => i)).toEqual([1, 2]);
  });

  it("should export as GeoJSON", () => {
    expect(branch.toGeoJSON()).toStrictEqual(
      branch.toGeoJSON("FeatureCollection")
    );
    expect(branch.toGeoJSON("FeatureCollection")).toMatchSnapshot();

    expect(branch.toGeoJSON("MultiLineString")).toStrictEqual(
      branch.toGeoJSON("Feature")
    );
    expect(branch.toGeoJSON("Feature")).toMatchSnapshot();

    expect(branch.toGeoJSON("MultiPolygon")).toMatchSnapshot();
  });

  it("should export as JSON", () => {
    expect(numbered.toJSON()).toMatchSnapshot();
    expect(branch.toJSON()).toMatchSnapshot();
  });
  it("should import as JSON", () => {
    expect(Branch.from(empty)).toBe(empty);

    expect(Branch.from(branch.toJSON(), branch.member)).toMatchObject(branch);
    expect(Branch.from(numbered.toJSON(), Number)).toMatchObject(numbered);

    // eslint-disable-next-line no-unused-vars
    let { member, ...almost } = branch;
    expect(Branch.from(branch.toGeoJSON())).toMatchObject(almost);
    expect(Branch.from(branch.toJSON())).toMatchObject(almost);
  });
});
describe("journey", () => {
  let emptyJourney, fullJourney;
  beforeEach(() => {
    emptyJourney = new Journey();

    fullJourney = new Journey(
      {
        name: "test flight"
      },
      new Branch(
        Location,
        {},
        new Location([0, -1, 1], { timestamp: 4000 }),
        new Location([1, 1, 0], { timestamp: 6000 })
      ),
      new Branch(
        Location,
        {},
        new Location([-1, 0, 0], { timestamp: 1000 }),
        new Location([0, 1, 1], { timestamp: 2000 })
      )
    );
  });

  it("extends Model", () => {
    expect(Journey.prototype).toBeInstanceOf(Model);
  });

  it("is instantiated", () => {
    expect(emptyJourney).toHaveProperty("branches", []);
    expect(emptyJourney).toHaveProperty("properties", {});

    expect(fullJourney).toHaveProperty(
      "branches",
      expect.arrayContaining([expect.any(Branch), expect.any(Branch)])
    );
    expect(fullJourney).toHaveProperty("properties", { name: "test flight" });
  });

  it("hold geom, bbox & length", () => {
    expect(fullJourney.geom).toMatchSnapshot();
    expect(fullJourney.bbox).toMatchSnapshot();
    expect(fullJourney.length).toMatchSnapshot();
  });

  it("should behave a bit like an array", () => {
    let branches;

    branches = jest.spyOn(fullJourney.branches, "forEach");
    fullJourney.forEach(x => x);
    expect(branches).toHaveBeenCalled();

    branches = jest.spyOn(fullJourney.branches, "map");
    fullJourney.map(x => x);
    expect(branches).toHaveBeenCalled();

    branches = jest.spyOn(fullJourney.branches, "filter");
    fullJourney.filter(() => true);
    expect(branches).toHaveBeenCalled();
  });

  it("should push, shift from last branch", () => {
    expect(fullJourney.last()).toBe(
      fullJourney.branches[fullJourney.branches.length - 1]
    );
    let lastbranch;

    lastbranch = jest.spyOn(fullJourney.last(), "push");
    fullJourney.push(new Location([1, 2, 3]));
    lastbranch = jest.spyOn(fullJourney.last(), "shift");
    fullJourney.shift();
    expect(lastbranch).toHaveBeenCalled();
  });

  it("should remove a given branch", () => {
    expect(fullJourney.removeBranch(-1)).toBe(fullJourney);
    expect(fullJourney.branches).toHaveLength(1);
    expect(fullJourney).toHaveProperty(
      "branches",
      expect.arrayContaining([expect.any(Branch)])
    );
  });

  it("should export as GeoJSON", () => {
    expect(fullJourney.toGeoJSON()).toStrictEqual(
      fullJourney.toGeoJSON("FeatureCollection")
    );
    expect(fullJourney.toGeoJSON("FeatureCollection")).toMatchSnapshot();

    expect(fullJourney.toGeoJSON("MultiLineString")).toStrictEqual(
      fullJourney.toGeoJSON("Feature")
    );
    expect(fullJourney.toGeoJSON("Feature")).toMatchSnapshot();

    expect(fullJourney.toGeoJSON("MultiPolygon")).toMatchSnapshot();
  });

  it("should export as JSON", () => {
    expect(fullJourney.toJSON()).toMatchSnapshot();
  });

  it("should import as JSON", () => {
    expect(Journey.from(fullJourney)).toBe(fullJourney);
    expect(Journey.from(fullJourney.toJSON())).toStrictEqual(fullJourney);
  });
});
