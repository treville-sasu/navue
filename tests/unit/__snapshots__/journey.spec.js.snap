// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`branch hold, geom, bbox & length 1`] = `
Array [
  Array [
    0,
    0,
    1,
  ],
  Array [
    1,
    1,
    0,
  ],
  Array [
    -1,
    -1,
    2,
  ],
]
`;

exports[`branch hold, geom, bbox & length 2`] = `
Array [
  -1,
  -1,
  1,
  1,
]
`;

exports[`branch hold, geom, bbox & length 3`] = `
Object {
  "precision": 6,
  "type": "Distance",
  "unit": "m",
  "value": 471749,
}
`;

exports[`branch should export as GeoJSON 1`] = `
Object {
  "bbox": Array [
    -1,
    -1,
    1,
    1,
  ],
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          0,
          0,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": 1,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          1,
          0,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": 0,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          -1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": 0,
      },
      "type": "Feature",
    },
  ],
  "properties": Object {
    "a": 1,
  },
  "type": "FeatureCollection",
}
`;

exports[`branch should export as GeoJSON 2`] = `
Object {
  "bbox": Array [
    -1,
    -1,
    1,
    1,
  ],
  "geometry": Object {
    "coordinates": Array [
      Array [
        0,
        0,
        1,
      ],
      Array [
        1,
        1,
        0,
      ],
      Array [
        -1,
        -1,
        2,
      ],
    ],
    "type": "MultiLineString",
  },
  "properties": Object {
    "a": 1,
  },
  "type": "Feature",
}
`;

exports[`branch should export as GeoJSON 3`] = `
Object {
  "bbox": Array [
    -1,
    -1,
    1,
    1,
  ],
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            Array [
              1.000095,
              0.999905,
            ],
            Array [
              1.000095,
              1.000095,
            ],
            Array [
              0.999905,
              1.000095,
            ],
            Array [
              -0.000095,
              0.000095,
            ],
            Array [
              -0.000095,
              -0.000095,
            ],
            Array [
              0.000095,
              -0.000095,
            ],
            Array [
              1.000095,
              0.999905,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": Object {
        "altitude": 1,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            Array [
              -1.000095,
              -0.999905,
            ],
            Array [
              -1.000095,
              -1.000095,
            ],
            Array [
              -0.999905,
              -1.000095,
            ],
            Array [
              1.000095,
              0.999905,
            ],
            Array [
              1.000095,
              1.000095,
            ],
            Array [
              0.999905,
              1.000095,
            ],
            Array [
              -1.000095,
              -0.999905,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": Object {
        "altitude": 0,
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`branch should export as JSON 1`] = `
Object {
  "features": Array [
    1,
    2,
  ],
  "member": "Number",
  "properties": Object {
    "a": 1,
  },
  "type": "Branch",
}
`;

exports[`branch should export as JSON 2`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          0,
          0,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": 1,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          1,
          0,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": 0,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          -1,
          2,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": 0,
      },
      "type": "Feature",
    },
  ],
  "member": "Position",
  "properties": Object {
    "a": 1,
  },
  "type": "Branch",
}
`;

exports[`journey hold geom, bbox & length 1`] = `
Array [
  Array [
    Array [
      0,
      -1,
      1,
    ],
    Array [
      1,
      1,
      0,
    ],
  ],
  Array [
    Array [
      -1,
      0,
      0,
    ],
    Array [
      0,
      1,
      1,
    ],
  ],
]
`;

exports[`journey hold geom, bbox & length 2`] = `
Array [
  -1,
  -1,
  1,
  1,
]
`;

exports[`journey hold geom, bbox & length 3`] = `
Object {
  "precision": 6,
  "type": "Distance",
  "unit": "m",
  "value": 405887,
}
`;

exports[`journey should export as GeoJSON 1`] = `
Object {
  "bbox": Array [
    -1,
    -1,
    1,
    1,
  ],
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          0,
          -1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": Object {
          "precision": 1,
          "reference": undefined,
          "type": "Altitude",
          "unit": "m",
          "value": 1,
        },
        "timestamp": 4000,
      },
      "type": "Location",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          1,
          1,
          0,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": Object {
          "precision": 0,
          "reference": undefined,
          "type": "Altitude",
          "unit": "m",
          "value": 0,
        },
        "timestamp": 6000,
      },
      "type": "Location",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -1,
          0,
          0,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": Object {
          "precision": 0,
          "reference": undefined,
          "type": "Altitude",
          "unit": "m",
          "value": 0,
        },
        "timestamp": 1000,
      },
      "type": "Location",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          0,
          1,
          1,
        ],
        "type": "Point",
      },
      "properties": Object {
        "altitude": Object {
          "precision": 1,
          "reference": undefined,
          "type": "Altitude",
          "unit": "m",
          "value": 1,
        },
        "timestamp": 2000,
      },
      "type": "Location",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`journey should export as GeoJSON 2`] = `
Object {
  "bbox": Array [
    -1,
    -1,
    1,
    1,
  ],
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          0,
          -1,
          1,
        ],
        Array [
          1,
          1,
          0,
        ],
      ],
      Array [
        Array [
          -1,
          0,
          0,
        ],
        Array [
          0,
          1,
          1,
        ],
      ],
    ],
    "type": "MultiLineString",
  },
  "properties": Object {
    "name": "test flight",
  },
  "type": "Feature",
}
`;

exports[`journey should export as GeoJSON 3`] = `
Object {
  "bbox": Array [
    -1,
    -1,
    1,
    1,
  ],
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            Array [
              1.000121,
              0.99994,
            ],
            Array [
              1.00006,
              1.000121,
            ],
            Array [
              0.999879,
              1.00006,
            ],
            Array [
              -0.000121,
              -0.99994,
            ],
            Array [
              -0.00006,
              -1.000121,
            ],
            Array [
              0.000121,
              -1.00006,
            ],
            Array [
              1.000121,
              0.99994,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": Object {
        "altitude": Object {
          "precision": 1,
          "reference": undefined,
          "type": "Altitude",
          "unit": "m",
          "value": 1,
        },
        "timestamp": 4000,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            Array [
              0.000095,
              0.999905,
            ],
            Array [
              0.000095,
              1.000095,
            ],
            Array [
              -0.000095,
              1.000095,
            ],
            Array [
              -1.000095,
              0.000095,
            ],
            Array [
              -1.000095,
              -0.000095,
            ],
            Array [
              -0.999905,
              -0.000095,
            ],
            Array [
              0.000095,
              0.999905,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": Object {
        "altitude": Object {
          "precision": 0,
          "reference": undefined,
          "type": "Altitude",
          "unit": "m",
          "value": 0,
        },
        "timestamp": 1000,
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`journey should export as JSON 1`] = `
Object {
  "branches": Array [
    Object {
      "features": Array [
        Object {
          "geometry": Object {
            "coordinates": Array [
              0,
              -1,
              1,
            ],
            "type": "Point",
          },
          "properties": Object {
            "altitude": Object {
              "precision": 1,
              "reference": undefined,
              "type": "Altitude",
              "unit": "m",
              "value": 1,
            },
            "timestamp": 4000,
          },
          "type": "Location",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              1,
              1,
              0,
            ],
            "type": "Point",
          },
          "properties": Object {
            "altitude": Object {
              "precision": 0,
              "reference": undefined,
              "type": "Altitude",
              "unit": "m",
              "value": 0,
            },
            "timestamp": 6000,
          },
          "type": "Location",
        },
      ],
      "member": "Location",
      "properties": Object {},
      "type": "Branch",
    },
    Object {
      "features": Array [
        Object {
          "geometry": Object {
            "coordinates": Array [
              -1,
              0,
              0,
            ],
            "type": "Point",
          },
          "properties": Object {
            "altitude": Object {
              "precision": 0,
              "reference": undefined,
              "type": "Altitude",
              "unit": "m",
              "value": 0,
            },
            "timestamp": 1000,
          },
          "type": "Location",
        },
        Object {
          "geometry": Object {
            "coordinates": Array [
              0,
              1,
              1,
            ],
            "type": "Point",
          },
          "properties": Object {
            "altitude": Object {
              "precision": 1,
              "reference": undefined,
              "type": "Altitude",
              "unit": "m",
              "value": 1,
            },
            "timestamp": 2000,
          },
          "type": "Location",
        },
      ],
      "member": "Location",
      "properties": Object {},
      "type": "Branch",
    },
  ],
  "properties": Object {
    "name": "test flight",
  },
  "type": "Journey",
}
`;
